nimporter/nimporter.py

```diff
@@ -0,0 +1,280 @@
+
+
+def hash_changed(ext: ExtLib) -> bool:
+    if not ext.hash_filename.exists():
+        return True
+
+    return hash_extension(ext.relative_path) != ext.hash_filename.read_bytes()
+
+
+def should_compile(ext: ExtLib) -> bool:
+    return hash_changed(ext) or not ext.build_artifact.exists()
+
+
+def compile_extension_to_lib(ext: ExtLib) -> None:
+    if not should_compile(ext):
+        ic('Skipping', ext.full_path)
+        return
+
+    ic('Compiling', ext.full_path)
+
+    ensure_nimpy()
+
+    ext.pycache.mkdir(parents=True, exist_ok=True)
+
+    with convert_to_lib_if_needed(ext.full_path) as compilation_dir:
+        nim_module = compilation_dir / (ext.symbol + '.nim')
+        ic(nim_module)
+
+        with cd(compilation_dir) as tmp_cwd:
+            cli_args = ALWAYS_ARGS + [
+                # ! Nimporter decides the use of the C compiler that was used
+                # ! to build Python itself to prevent incompatibilities. This
+                # ! is similar to exporting where several C compilers are used.
+                f'--cc:{get_c_compiler_used_to_build_python()}',
+                nim_module.name
+            ]
+
+            ic(cli_args)
+
+            code, _, stderr = run_process(
+                cli_args,
+                'NIMPORTER_INSTRUMENT' in os.environ
+            )
+
+            if code:
+                raise CompilationFailedException(stderr)
+
+            # Remove Windows debugging symbols if using MSVC on Win32
+            for debug_ext in ['.exp', '.lib']:
+                debug_file = compilation_dir / (ext.symbol + debug_ext)
+                if debug_file.exists():
+                    ic(debug_file).unlink()
+
+        platform = get_host_info()[0]
+        find_ext = {WINDOWS: '.dll', MACOS: '.dylib', LINUX: '.so'}[platform]
+
+        # The only way the artifact wouldn't exist is if Nim failed to
+        # compile the library but didn't write to the standard error stream
+        # which is currently not how the Nim compiler behaves.
+        # This shouldn't fail.
+        (tmp_build_artifact,) = tmp_cwd.glob(f'*{find_ext}')
+
+        shutil.move(tmp_build_artifact, ext.build_artifact)
+
+        write_hash(ext)
+
+
+def validate_spec(spec: ModuleSpec) -> None:
+    """
+    Validates that a given Nim extension can be successfully imported.
+
+    Since a spec is used to import a module and the body of this function
+    attempts an import, successful imports will imply importing the module
+    twice. However, it's ok that the module will be imported twice because
+    it's better to have good error messages than to be fast in this case.
+
+    Args:
+        spec(Spec): the spec to validate if its module can be imported.
+
+    Raises:
+        A NimporterException if the spec cannot be used to import the given
+        Nim module/library.
+    """
+    try:
+        util.module_from_spec(spec)
+    except ImportError:
+        msg = '''
```

Comment:
We should consider creating a file text.py where we put this kind of text messages to declutter the module.